connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java

```diff
@@ -132,25 +132,36 @@ public DelegatingClassLoader(List<String> pluginPaths) {
         return connectorClientConfigPolicies;
     }
 
+    public PluginClassLoader pluginClassLoader(String name) {
+        if (PluginUtils.shouldLoadInIsolation(name)) {
+            SortedMap<PluginDesc<?>, ClassLoader> inner = pluginLoaders.get(name);
+            if (inner != null) {
+                ClassLoader pluginLoader = inner.get(inner.lastKey());
+                if (pluginLoader instanceof PluginClassLoader)  {
```

Comment:
This check is interesting. Based on the plugin loader-specific error message in `connectorLoader` it should have been included before this PR, so it seems fine to add here. However, just out of curiosity, is this strictly necessary for the bug fix, and if so, what breaks if it's not included?

Reply:
It's actually possible for pluginLoaders to include a non-`PluginClassLoader`:
https://github.com/apache/kafka/blob/trunk/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java#L192-L196

This was a lateral refactor, and this was originally here: https://github.com/apache/kafka/blob/trunk/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java#L370

Reply:
It appears that I have changed the behavior with this refactor, but as far as I can tell, the only change is that when before this could have returned `getParent()`, it now returns itself.

This would cause CP connectors to use the delegating classloader instead of the system classloader, and go through one extra data structure lookup on each class load.
It would also make storing `getParent()` in the plugin classloaders map meaningless, where before it "cached" the delegation of CP connectors.

Is it ever correct for the delegating classloader to be the "connector loader"? Shouldn't it always be either the system classloader or the plugin classloader? Perhaps this function should return `getParent()` instead of `this`. This makes the "caching" behavior of storing `getParent()` doubly useless.

Maybe I should revert this refactor, it was never really necessary for the functionality, and it might be better to duplicate the `pluginLoaders` map lookup logic since it's really straightforward anyway.

Reply:
I believe the existing behavior may be intentional and have something to do with the parent of the `DelegatingClassLoader` being something other than the system class loader. However, I'm not familiar enough with this part of the code base to say for sure; we should probably err on the side of keeping the existing functionality as-is unless someone else (e.g., @kkonstantine or a committer) can vouch that the current changes here wouldn't break anything.